polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -551,6 +555,75 @@ where
 	Some(processed_code_hashes)
 }
 
+async fn update_active_leaves<Sender>(
+	sender: &mut Sender,
+	mut validation_backend: impl ValidationBackend,
+	update: ActiveLeavesUpdate,
+) where
+	Sender: SubsystemSender<ChainApiMessage> + SubsystemSender<RuntimeApiMessage>,
+{
+	let ancestors = if let Some(ref leaf) = update.activated {
+		get_block_ancestors(sender, leaf.hash).await
+	} else {
+		None
+	};
+	if let Err(err) = validation_backend.update_active_leaves(update, ancestors).await {
+		gum::warn!(
+			target: LOG_TARGET,
+			?err,
+			"cannot update active leaves in validation backend",
+		);
+	};
+}
+
+async fn get_allowed_ancestry_len<Sender>(sender: &mut Sender, relay_parent: Hash) -> Option<usize>
+where
+	Sender: SubsystemSender<ChainApiMessage> + SubsystemSender<RuntimeApiMessage>,
+{
+	match prospective_parachains_mode(sender, relay_parent).await {
+		Ok(ProspectiveParachainsMode::Enabled { allowed_ancestry_len, .. }) =>
+			Some(allowed_ancestry_len),
+		Ok(_) => {
+			gum::warn!(target: LOG_TARGET, "async backing is disabled");
+			None
+		},
+		Err(err) => {
+			gum::warn!(target: LOG_TARGET, ?err, "cannot request prospective parachains mode");
+			None
+		},
+	}
+}
+
+async fn get_block_ancestors<Sender>(sender: &mut Sender, relay_parent: Hash) -> Option<Vec<Hash>>
+where
+	Sender: SubsystemSender<ChainApiMessage> + SubsystemSender<RuntimeApiMessage>,
+{
+	let Some(allowed_ancestry_len) = get_allowed_ancestry_len(sender, relay_parent).await else {
+		return None
+	};
+
+	let (tx, rx) = oneshot::channel();
+	sender
+		.send_message(ChainApiMessage::Ancestors {
+			hash: relay_parent,
+			k: allowed_ancestry_len + 1,
```

Comment:
Why do we need `allowed_ancestry_len + 1` ? The tip of chain (`relay_parent`) of the chain is the additional +1 and that is already stored separately in `ActiveLeafWithAncestors`.